### ISSUE TITLE
Kernel types for polymorphic objects

### ISSUE THREAD

**Issue by ejgallego** [rocq-prover/rocq#6365]
The more I look into it, the more I have problems managing the distinction between polymorphic objects and those that are not.

In particular, I am starting to wonder if types such as `mutual_inductive_body` should be really split/tagged into 3 different types depending their "universe" status, and thus all testing in the style of `inductive_is_cumulative` should go away.

A similar reasoning does apply to `definition_entry`, which however already carries a parameter but in this case for the wrong reasons [due to "side-effects"].

What do you think folks? Should such a patch be tried. @ppedrot , have you already proposed something like this?

**Comment by ppedrot** [comment]
I already have a branch that does this, using an additional type parameter and GADTs to statically enforce the polymorphic status of definitions and inductive types. It's kind of unwieldy though.

**Comment by ejgallego** [comment]
Interesting, it'd be great to have a look at it. I was [naively] thinking that most codepaths should be quite uniform among the several kinds of definitions, but it turns out that they are maybe not!

**Comment by ejgallego** [comment]
@ppedrot should we assign this to you? I get the feeling that you are getting closer.

**Comment by ejgallego** [comment]
If you drop a small TODO we could surely help.

**Comment by gares** [comment]
While I understand why we keep polymorphic stuff separate, I frankly don't get where we want to go in the long term. In particular, if the calculus supports polymorphism, then we should try to converge to a single status. Eg, today `list` and `nat`, `map` and `addn`, ... are of the same kind even if the first of the two pairs is polymorphic while the other is not. Sure, this is universe polymorphism and not type polymorphism, but I don't see where the fundamental difference lies since no-quantification really means monomorphic to me. I would find more natural to have monorphic stuff be just paired with empty universe stuff. Am I dreaming?

**Comment by gares** [comment]
It goes without saying that I also think the code paths should converse, as they do for type polymorphic stuff.

**Comment by SkySkimmer** [comment]
> I would find more natural to have monorphic stuff be just paired with empty universe stuff. 

I think @ppedrot is working in that direction. IIRC the main issue is disentangling the way side effects manage monomorphic universes.

**Comment by ppedrot** [comment]
@SkySkimmer Yeah, that's a quite accurate summary of the current mess and way out.
